mesonbuild/rewriter.py

```diff
@@ -23,36 +23,301 @@
+                if choices is not None:
+                    assert(isinstance(choices, list))
+                    if cmd[key] not in choices:
+                        raise RewriterException('Invalid value of "{}": Possible values are {} but provided was "{}"'
+                                                .format(key, choices, cmd[key]))
+            return f(*wrapped_args, **wrapped_kwargs)
+
+        return wrapped
+
+rewriter_keys = {
+    'target': {
+        'target': (str, None, None),
+        'operation': (str, None, ['src_add', 'src_rm', 'test']),
+        'sources': (list, [], None),
+        'debug': (bool, False, None)
+    }
+}
+
+class Rewriter:
+    def __init__(self, sourcedir: str, generator: str = 'ninja'):
+        self.sourcedir = sourcedir
+        self.interpreter = IntrospectionInterpreter(sourcedir, '', generator)
+        self.id_generator = AstIDGenerator()
+        self.modefied_nodes = []
+        self.functions = {
+            'target': self.process_target,
+        }
+
+    def analyze_meson(self):
+        mlog.log('Analyzing meson file:', mlog.bold(os.path.join(self.sourcedir, environment.build_filename)))
+        self.interpreter.analyze()
+        mlog.log('  -- Project:', mlog.bold(self.interpreter.project_data['descriptive_name']))
+        mlog.log('  -- Version:', mlog.cyan(self.interpreter.project_data['version']))
+        self.interpreter.ast.accept(AstIndentationGenerator())
+        self.interpreter.ast.accept(self.id_generator)
+
+    def find_target(self, target: str):
+        for i in self.interpreter.targets:
+            if target == i['name'] or target == i['id']:
+                return i
+        return None
+
+    @RequiredKeys(rewriter_keys['target'])
+    def process_target(self, cmd):
+        mlog.log('Processing target', mlog.bold(cmd['target']), 'operation', mlog.cyan(cmd['operation']))
+        target = self.find_target(cmd['target'])
+        if target is None:
+            mlog.error('Unknown target "{}" --> skipping'.format(cmd['target']))
+            if cmd['debug']:
+                pprint(self.interpreter.targets)
+            return
+        if cmd['debug']:
+            pprint(target)
+
+        # Utility function to get a list of the sources from a node
+        def arg_list_from_node(n):
+            args = []
+            if isinstance(n, mparser.FunctionNode):
+                args = list(n.args.arguments)
+                if n.func_name in build_target_functions:
+                    args.pop(0)
+            elif isinstance(n, mparser.ArrayNode):
+                args = n.args.arguments
+            elif isinstance(n, mparser.ArgumentNode):
+                args = n.arguments
+            return args
+
+        if cmd['operation'] == 'src_add':
+            #################################
+            ### Add sources to the target ###
+            #################################
```

Comment:
Don't do this. There is no reason to have shouty comment syntax.

Reply:
Sure, removed with the last commit.